compiler/noirc_evaluator/src/ssa/opt/basic_conditional.rs

```diff
@@ -0,0 +1,379 @@
+    block: BasicBlockId,
+    cfg: &ControlFlowGraph,
+    function: &Function,
+) -> Option<BasicConditional> {
+    // jump overhead is the cost for doing the conditional and jump around the blocks
+    // I use 10 as a rough estimate, real cost is less.
+    let jump_overhead = 10;
+    let mut successors = cfg.successors(block);
+    let mut result = None;
+    // a conditional must have 2 branches
+    if successors.len() == 2 {
+        let left = successors.next().unwrap();
+        let right = successors.next().unwrap();
+        let mut left_successors = cfg.successors(left);
+        let mut right_successors = cfg.successors(right);
+        let left_successors_len = left_successors.len();
+        let right_successors_len = right_successors.len();
+        let next_left = left_successors.next();
+        let next_right = right_successors.next();
+        if next_left == Some(block) || next_right == Some(block) {
+            // this is a loop, not a conditional
+            return None;
+        }
+        if left_successors_len == 1 && right_successors_len == 1 && next_left == next_right {
+            // The branches join on one block so it is a non-nested conditional
+            let cost_left = block_cost(left, &function.dfg);
+            let cost_right = block_cost(right, &function.dfg);
+            // For the flattening to be valuable, we compare the cost of the flattened code with the average cost of the 2 branches,
+            // including an overhead to take into account the jumps between the blocks.
+            let cost = cost_right.saturating_add(cost_left);
+            if cost < cost / 2 + jump_overhead {
+                if let Some(TerminatorInstruction::JmpIf {
+                    condition: _,
+                    then_destination,
+                    else_destination,
+                    call_stack: _,
+                }) = function.dfg[block].terminator()
+                {
+                    result = Some(BasicConditional {
+                        block_entry: block,
+                        block_then: Some(*then_destination),
+                        block_else: Some(*else_destination),
+                        block_exit: next_left.unwrap(),
+                    });
+                }
+            }
+        } else if left_successors_len == 1 && next_left == Some(right) {
+            // Left branch joins the right branch, it is a if/then statement with no else
+            // I am not sure whether this case can happen, but it is not difficult to handle it
+            let cost = block_cost(left, &function.dfg);
+            if cost < cost / 2 + jump_overhead {
+                if let Some(TerminatorInstruction::JmpIf {
+                    condition: _,
+                    then_destination,
+                    else_destination,
+                    call_stack: _,
+                }) = function.dfg[block].terminator()
+                {
+                    if left == *then_destination {
```

Comment:
I think this would be more readable if we did something like:
```rust
let (block_then, block_else) = if left == *then_destination {
   (Some(left), None)
} else if left == *else_destination {
   (None, Some(left))
};
result = Some(BasicConditional {
       block_entry: block,
       block_then,
       block_else,
       block_exit: right,
});
```

Reply:
Done